UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.IO;
+using System.IO.Abstractions;
+using FluentAssertions;
+using GitCommands;
+using GitUIPluginInterfaces;
+using NSubstitute;
+using NUnit.Framework;
+
+namespace GitCommandsTests
+{
+    [TestFixture]
+    public class CommitTemplateManagerTests
+    {
+        private IGitModule _module;
+        private FileBase _file;
+        private IFileSystem _fileSystem;
+        private CommitTemplateManager _manager;
+
+
+        [SetUp]
+        public void Setup()
+        {
+            _module = Substitute.For<IGitModule>();
+            _file = Substitute.For<FileBase>();
+            _fileSystem = Substitute.For<IFileSystem>();
+            _fileSystem.File.Returns(_file);
+            _manager = new CommitTemplateManager(_module, _fileSystem);
+        }
+
+
+        [TestCase(null)]
+        [TestCase("")]
+        public void LoadGitCommitTemplate_commit_template_not_defined(string template)
+        {
+            _module.GetEffectiveSetting("commit.template").Returns(template);
+
+            _manager.LoadGitCommitTemplate().Should().BeNull();
+        }
+
+        [TestCase("~/template.txt")]
+        [TestCase("./template.txt")]
+        [TestCase("template.txt")]
+        public void LoadGitCommitTemplate_path_not_rooted(string template)
+        {
+            const string workingDir = @"c:\dev\repo";
+            _module.WorkingDir.Returns(workingDir);
+            _module.GetEffectiveSetting("commit.template").Returns(template);
+
+            ((Action)(() => _manager.LoadGitCommitTemplate())).ShouldThrow<FileNotFoundException>();
+        }
+
+        [Test]
+        public void LoadGitCommitTemplate_should_load_file_content()
+        {
+            const string workingDir = @"c:\dev\repo";
+            _module.WorkingDir.Returns(workingDir);
+            _module.GetEffectiveSetting("commit.template").Returns("./template.txt");
+            _file.Exists(Arg.Any<string>()).Returns(true);
+            _file.ReadAllText(Arg.Any<string>()).Returns("line1\r\nline2\rline3\nline4");
```

Comment:
If I change the controller implementation to:
```
string commitTemplate = _fileSystem.File.ReadAllText(fileName, _module.CommitEncoding).Replace("\r", "");
```
then the test will fail.

Reply:
and it must - you've just changed the behavior.

Reply:
I think that we are not able to find a compromise here. When tests are written to test the implementation, then yes, it must. But when they are written to verify assumptions (pre and post conditions like given-when-then) then it should not.

Reply:
If in some point in time we decided to replace
`string fileName = _module.GetEffectiveSetting("commit.template");`
with 
```
if (!_module.EffectiveConfigFile.HasValue("commit.template"))
  return;
string fileName = _module.EffectiveConfigFile.GetValue("commit.template");
```
then again the test will fail.

Reply:
yes, that is correct - you have changed the contract

Reply:
How is that? The LoadGitCommitTemplate_commit_template_not_defined still passes. The two other tests don't.

Reply:
> But when they are written to verify assumptions (pre and post conditions like given-when-then) then it should not.

You have to arrange before the assertion. The arrangement is based on an existing contract.

How would you test it?

Reply:
Create a config file (in memory or on hard disk) that contains `commit.template` entry.

Reply:
The way I see it - any tests are better than no tests at all. 
Currently the test coverage is very limited and academic discussions how a very unlikely change in a possible future may break a test is counter productive.

If you modify a contract any functionality dependent on that contract will be affected and there is a chance of regression. Of course you will need to test it. 
I would be very worried if tests wouldn't fail in such case.

Reply:
> Create a config file (in memory or on hard disk) that contains commit.template entry.

How would it change things?

Also this will not be just testing the `CommitTemplateManager` but also `GitModule` reading settings, then `FileSystem` checking and loading files.
This is wrong

Reply:
Perhaps we have to define the meaning of `contract` more precisely as I don't see any change in the contract here.

>The way I see it - any tests are better than no tests at all.

Sure. But the discussion is at the general level and it should happen in https://github.com/gitextensions/gitextensions/pull/3898. The intention is to increase the test coverage and the discussion is on how to do it properly (as we are at the beginning of that process). 
I know that the change I gave as an example is unlikely to happen. It was only to illustrate how fragile tests written in this manner are.

Reply:
>How would it change things?

I am not able to explain it more. Go implement and you will see.

>This is wrong

As you see it is disputable. Such an approach has an added value. You cross-verify that the assumptions you made on implementation and external interfaces are good. Read https://medium.com/python-pandemonium/mock-or-not-to-mock-41965d33f175 as an example.

Reply:
An interesting read, but I didn't find it convincing

Reply:
I feel like I am wasting my time. I can not reason against `I didn't find it convincing`
You didn't answer to `How is that?` but then you continue conversation like it was agreed that I changed the contract: `If you modify a contract any functionality dependent on that contract will be affected and there is a chance of regression.`